### DISCUSSION TITLE
Abstract transaction layer and manual contorl

### DISCUSSION THREAD

**Discussion by mr-scrpt** [prisma/prisma#19903] [Q&A] [answered]
Hello. In my NestJs application, I have an abstraction layer - the repository layer, where the application interacts with Prisma. This is done so that the services, which contain the business logic of the application, are not tied to a specific ORM implementation, but work with the repository abstraction. Everything was working fine until I needed to work with transactions. Let's say I need to create an article and its associated tags. I have two different methods in two different repositories, and I need to call both of them within a single transaction, which needs to be done in the service. Here's a simplified example (article.service.ts):
```
async createArticle(
  articleToDB: ArticleToDBDto,
  tagList: string[],
): Promise<void> {
  await this.articleRepository.createArticle(articleToDB);
  await this.tagRepository.createTagList(tagList);
}

```
I thought I could solve this issue by adding an additional wrapper for transaction management, and I found a library that does what I need: [https://github.com/mokuteki225/propagated-transactions](https://github.com/mokuteki225/propagated-transactions). However, the problem is that I couldn't find any transaction management methods in Prisma. So I can't do something like this:
```
const PrismaTransactionRunner = {
  start: async () => {
    const trx = await prisma.$transaction();
    return trx;
  },
  commit: async (trx) => {
    return trx.$commit();
  },
  rollback: async (trx) => {
    return trx.$rollback();
  },
};
```
And I can't do something like this either:

```
async createArticle(
  articleToDB: ArticleToDBDto,
  tagList: string[],
): Promise<void> {
  const callback = async () => {
    try {
      await this.articleRepository.createArticle(articleToDB);
      await this.tagRepository.createTagList(tagList);
      await this.ptx.commit();
    } catch (err) {
      await this.ptx.rollback();
    }
  }
}

```
Can you please advise if there's a way to achieve what I have in mind using Prisma? Perhaps there are some approaches to solve my problem?

**Comment by nurul3101** [comment, marked as answer]
Hello @mr-scrpt 👋 

I think you are looking for [Prisma Client Extensions](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions), specifically this extension: [callback-free-itx](https://github.com/prisma/prisma-client-extensions/tree/main/callback-free-itx).

Can you have a look and check if that helps?

**Reply by mr-scrpt** [reply]
Thanks, this looks like what I need
